services/autorust/codegen/src/crates.rs

```diff
@@ -0,0 +1,72 @@
+use camino::{Utf8Path, Utf8PathBuf};
+use serde::Deserialize;
+use std::io::BufRead;
+use std::{
+    fs::{self, File},
+    io::BufReader,
+    str::FromStr,
+};
+pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+/// Get all directories below the given directory.
+fn list_dirs_in(dir: impl AsRef<Utf8Path>) -> Result<Vec<Utf8PathBuf>> {
+    let mut dirs = Vec::new();
+    let paths = fs::read_dir(dir.as_ref())?;
+    for path in paths.flatten() {
+        if let Some(path) = Utf8Path::from_path(&path.path()) {
+            if path.is_dir() && path.join("Cargo.toml").exists() {
+                dirs.push(path.to_path_buf());
+            }
+        }
+    }
+    Ok(dirs)
+}
+
+pub fn list_dirs() -> Result<Vec<Utf8PathBuf>> {
+    let mut names: Vec<_> = list_dirs_in("../mgmt")?.into_iter().collect();
+    names.extend(list_dirs_in("../svc")?.into_iter());
+    names.sort();
+    Ok(names)
+}
+
+pub fn list_crate_names() -> Result<Vec<String>> {
+    let mut names: Vec<_> = list_dirs_in("../mgmt")?
+        .into_iter()
+        .filter_map(|d| d.file_name().map(|d| format!("azure_mgmt_{}", d)))
+        .collect();
+    names.extend(
+        list_dirs_in("../svc")?
+            .into_iter()
+            .filter_map(|d| d.file_name().map(|d| format!("azure_svc_{}", d))),
+    );
+    names.sort();
+    Ok(names)
+}
+
+pub fn has_version(name: &str, version: &str) -> Result<bool> {
+    Ok(get_versions(name)?.iter().any(|v| v.vers.as_str() == version))
+}
+
+/// Gets all the versions for a given crate
+/// Expects https://github.com/rust-lang/crates.io-index to be cloned as a sibling directory.
+fn get_versions(crate_name: &str) -> Result<Vec<CrateVersion>> {
+    // all of these crates begin with "azure".
+    let path = format!("../../../crates.io-index/az/ur/{}", crate_name);
```

Comment:
Nevermind, I figured out how this is used and it's reasonable.  Though, we may want to do a shallow checkout at runtime rather than rely on this being up-to-date by the user.

Reply:
This utility function is only used in the examples where I am auditing published crates. I prefer sibling projects over git submodules.

Reply:
I'm not suggesting submodules.  I'm suggesting something to the effect of creating a temp dir, doing a `git checkout --depth 1`, getting the data you need, then cleaning the temp dir.

Both submodules and sibling projects rely on manual user interaction, which we should avoid.